### PR TITLE
Close Filesystem on exit with Minion Tasks

### PR DESCRIPTION
**Pull request by KKcorps**
When Filesystems are created outside the factory, we should ensure that they are closed on exit. This doesn't happen a lot of times in minion tasks and leads to memory leaks such as this - 
https://coder-shankar.medium.com/aws-sdk-memory-leak-63e23dcdc6f0

### CODE REVIEW

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskUtils.java [hunk 1]
```diff
@@ -50,7 +49,7 @@ static PinotFS getInputPinotFS(Map<String, String> taskConfigs, URI fileURI)
       return pinotFS;
     }
     // Fallback to use the PinotFS created by Minion Server configs
-    return PinotFSFactory.create(fileURIScheme);
+    return null;
```
Comment (jackjlli) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskUtils.java:52]: Should we update the comment 1 line above?

Comment (KKcorps) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskUtils.java:51]: Thanks for pointing out! Removed it now.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskExecutor.java [hunk 2]
```diff
@@ -177,12 +178,24 @@ private void pushSegment(String tableName, Map<String, String> taskConfigs, URI
       outputSegmentDirURI = URI.create(taskConfigs.get(BatchConfigProperties.OUTPUT_SEGMENT_DIR_URI));
     }
     PinotFS outputFileFS = SegmentGenerationAndPushTaskUtils.getOutputPinotFS(taskConfigs, outputSegmentDirURI);
+    boolean closeOutputFSOnExit = true;
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskExecutor.java:181]: It seems a bit error prone to me to use this flag to tell if the fs is from a cache of a factory, e.g. the getOutputPinotFS () can return LOCAL_PINOT_FS and it is closed (although it's a noop). 

How about always return a new instance from getOutputPinotFS(), so it should be closed all the time after use. It incurs some overhead but less error prone, and it's for task generator/executor (batching processing) so the overhead can be tolerable.

**Review by Jackie-Jiang — COMMENTED** [review]
Good catch on the resource leak!

+1 with @klsince on always closing the `PinotFS`, and use try-with-resource to make it more robust and readable. I feel the overhead should be okay to always create a new `PinotFS` since it is not in performance critical path. If that becomes a bottleneck in the future, we can also add a `PinotFS` delegate class which does not close the underlying `PinotFS`

**Comment by KKcorps** [comment]
I can do that. But then, the question is what should one do if no FS configs are mentioned in the task config. Currently we get the FS from cache which uses minion FS configs.

**Comment by walterddr** [comment]
another thought since we are talking about how to map fsClassName to pinotFS. 

should we also address the fact that users might have different `fsConfiguration` for the same `fsClassName`? for example 2 S3PinotFS (same scheme) but different configuration for access keys. i don't think the current PinotFSFactory handles this

**Comment by KKcorps** [comment]
@walterddr that is out of the scope for this. In minion tasks, user provided FS configs take priority over cached FS.

**Comment by walterddr** [comment]
thanks for the clarification. originally i thought you will change the key of the cached map but seems like it only changes the content which seems fine

File: pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/PinotFSFactory.java [hunk 3]
```diff
@@ -74,9 +84,21 @@ public static void init(PinotConfiguration fsFactoryConfig) {
...
-    Preconditions.checkState(pinotFS != null, "PinotFS for scheme: %s has not been initialized", scheme);
-    return pinotFS;
+    PinotFSInfo pinotFSInfo = PINOT_FS_MAP.get(scheme);
+    Preconditions.checkState(pinotFSInfo != null, "PinotFS for scheme: %s has not been initialized", scheme);
+    return pinotFSInfo.getPinotFS();
+  }
+
+  public static PinotFS createNewInstance(String scheme) {
+    PinotFSInfo pinotFSInfo = PINOT_FS_MAP.get(scheme);
+    Preconditions.checkState(pinotFSInfo != null, "PinotFS for scheme: %s has not been initialized", scheme);
```
Comment (klsince) [inline pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/PinotFSFactory.java:94]: perhaps make the error msg "FS scheme: %s has not been `registered`" to be more specific

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskUtils.java [hunk 4]
```diff
@@ -67,8 +66,7 @@ static PinotFS getOutputPinotFS(Map<String, String> taskConfigs, URI fileURI)
       pinotFS.init(fsProps);
       return pinotFS;
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentgenerationandpush/SegmentGenerationAndPushTaskUtils.java:67]: how about add a public method PinotFSFactory.createNewInstance(fileURIScheme, fsClass, fsProps) or make PinotFSFactory. createPinotFSInstance() public and call it here?

File: pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/PinotFSFactory.java [hunk 5]
```diff
@@ -74,9 +84,21 @@ public static void init(PinotConfiguration fsFactoryConfig) {
   }
 
   public static PinotFS create(String scheme) {
```
[the remark this answers is not in the document]
Comment (KKcorps) [reply pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/PinotFSFactory.java:86]: Yes, this is the reason I didn't rename.

File: pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/PinotFSFactory.java [hunk 6]
```diff
@@ -39,24 +39,34 @@ private PinotFSFactory() {
...
+      put(LOCAL_PINOT_FS_SCHEME, new PinotFSInfo(null, null, new LocalPinotFS()));
     }
   };
 
   public static void register(String scheme, String fsClassName, PinotConfiguration fsConfiguration) {
     try {
-      LOGGER.info("Initializing PinotFS for scheme {}, classname {}", scheme, fsClassName);
-      PinotFS pinotFS = PluginManager.get().createInstance(fsClassName);
-      pinotFS.init(fsConfiguration);
-      PINOT_FS_MAP.put(scheme, pinotFS);
```
[the remark this answers is not in the document]
Comment (KKcorps) [reply pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/PinotFSFactory.java:53]: You're right. This approach seems cleaner and much-less error-prone for caller. Made the changes.

File: pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/NoClosePinotFS.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.List;
+import org.apache.pinot.spi.env.PinotConfiguration;
+
+
+public class NoClosePinotFS implements PinotFS {
```
Comment (klsince) [inline pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/NoClosePinotFS.java:29]: nit: maybe make this private to this pkg, as not sure if it'd be used outside.

**Review by klsince — APPROVED** [review]
👍

File: pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/NoClosePinotFS.java (hunk 7 above)
Comment (KKcorps) [reply pinot-spi/src/main/java/org/apache/pinot/spi/filesystem/NoClosePinotFS.java:29]: done

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**